shell/platform/linux/fl_text_input_plugin.cc

```diff
@@ -310,10 +322,12 @@ static void im_commit_cb(FlTextInputPlugin* self, const gchar* text) {
 static void im_preedit_end_cb(FlTextInputPlugin* self) {
   FlTextInputPluginPrivate* priv = static_cast<FlTextInputPluginPrivate*>(
       fl_text_input_plugin_get_instance_private(self));
+  std::string text_before_change = priv->text_model->GetText();
   priv->text_model->EndComposing();
   if (priv->enable_delta_model) {
     flutter::TextEditingDelta delta = flutter::TextEditingDelta(
-        "", flutter::TextRange(-1, -1), priv->text_model->GetText());
+        text_before_change, flutter::TextRange(-1, -1),
```

Comment:
I'm surprised this works. The type of the base/extent positions are `size_t`, which is unsigned.

Reply:
Here's a prototype of something we could use to do this safely:
https://github.com/flutter/engine/compare/main...cbracken:flutter_engine:add-invalid-range?expand=1

If this seems useful, I could commit it, but when I added TextRange originally my thought was to see how far along we could make it without modelling things using invalid positions like `-1` since some platforms (macOS, iOS) don't model any positions with negative numbers, so we're going to need to do translation between models somewhere regardless and TextRange was cleaner without the concept of negative values.

Reply:
I think in this case it may be correct to use composing_before_change. So this may not be needed. Thanks for checking that out!

Reply:
correction, a delta range of `(-1,-1)`, is needed but we should just use `TextEditingDelta(const std::u16string & text)` for this case.